infer/environment.cc

```diff
@@ -1334,10 +1334,22 @@ core::TypePtr Environment::processBinding(core::Context ctx, const cfg::CFG &inW
             },
             [&](cfg::LoadSelf &l) {
                 ENFORCE(l.link);
-                if (l.link->result->main.blockSpec.rebind.exists()) {
-                    tp.type = l.link->result->main.blockSpec.rebind.data(ctx)->externalType();
-                    tp.origins.emplace_back(core::Loc(ctx.file, bind.loc));
+                auto rebind = l.link->result->main.blockSpec.rebind;
+
+                if (rebind) {
+                    if (core::isa_type<core::SelfType>(rebind)) {
+                        tp.type = l.link->result->main.receiver;
+                    } else if (core::isa_type<core::LambdaParam>(rebind)) {
+                        tp.type = core::cast_type<core::AppliedType>(l.link->result->main.receiver)
+                                      ->klass.data(ctx)
+                                      ->attachedClass(ctx)
+                                      .data(ctx)
+                                      ->resultType;
```

Comment:
This is not the same as `T.attached_class`, which relies on the defaulting behavior for the upper bound of a `type_template` variable. The upper bound in that case will be computed by `externalType`, which is slightly different.